[15/n] [torch/elastic] Introduce _RendezvousStateHolder

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57151 [23/n] [torch/elastic] Introduce the implementation of DynamicRendezvousHandler
* #57150 [22/n] [torch/elastic] Introduce a new from_backend static constructor for DynamicRendezvousHandler
* #57149 [21/n] [torch/elastic] Introduce _RendezvousJoinOp
* #57148 [20/n] [torch/elastic] Introduce _RendezvousExitOp
* #57147 [19/n] [torch/elastic] Introduce _RendezvousKeepAliveOp
* #57146 [18/n] [torch/elastic] Introduce _RendezvousCloseOp
* #57145 [17/n] [torch/elastic] Introduce _DistributedRendezvousOpExecutor
* #57144 [16/n] [torch/elastic] Introduce _RendezvousOpExecutor
* **#56538 [15/n] [torch/elastic] Introduce _RendezvousStateHolder**
* #57143 [14/n] [torch/elastic] Introduce a name attribute to _PeriodicTimer
* #57142 [13/n] [torch/elastic] Extend the return type of RendezvousBackend's set_state method
* #57141 [12/n] [torch/elastic] Rename last_keep_alives to last_heartbeats in _RendezvousState
* #57140 [11/n] [torch/elastic] Add heartbeat timeout to RendezvousTimeout
* #57139 [10/n] [torch/elastic] Add comparison operators to _NodeDesc
* #56537 [9/n] [torch/elastic] Introduce RendezvousSettings
* #56536 [8/n] [torch/elastic] Add unit tests for _RendezvousState
* #56535 [7/n] [torch/elastic] Rename _Rendezvous to _RendezvousState
* #56534 [6/n] [torch/elastic] Reorder type definitions in dynamic_rendezvous.py
* #56533 [5/n] [torch/elastic] Introduce the delay utility function
* #56532 [4/n] [torch/elastic] Fix the finalizer of PeriodicTimer

This PR introduces the `_RendezvousStateHolder` type that is responsible for synchronizing the local rendezvous state via the backend with the other nodes in the job.

Differential Revision: [D27892600](https://our.internmc.facebook.com/intern/diff/D27892600/)

cbalioglu committed:
[10/n] [torch/elastic] Introduce _RendezvousStateHolder

This PR introduces the `_RendezvousStateHolder` type that is responsible for synchronizing the local rendezvous state via the backend with the other nodes in the job.

Differential Revision: [D27892600](https://our.internmc.facebook.com/intern/diff/D27892600/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[10/n] [torch/elastic] Introduce _RendezvousStateHolder"


This PR introduces the `_RendezvousStateHolder` type that is responsible for synchronizing the local rendezvous state via the backend with the other nodes in the job.

Differential Revision: [D27892600](https://our.internmc.facebook.com/intern/diff/D27892600/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[10/n] [torch/elastic] Introduce _RendezvousStateHolder"


This PR introduces the `_RendezvousStateHolder` type that is responsible for synchronizing the local rendezvous state via the backend with the other nodes in the job.

Differential Revision: [D27892600](https://our.internmc.facebook.com/intern/diff/D27892600/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[10/n] [torch/elastic] Introduce _RendezvousStateHolder"


This PR introduces the `_RendezvousStateHolder` type that is responsible for synchronizing the local rendezvous state via the backend with the other nodes in the job.

Differential Revision: [D27892600](https://our.internmc.facebook.com/intern/diff/D27892600/)

[ghstack-poisoned]

tierex commented on torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:
```diff
@@ -243,6 +245,129 @@ def __init__(self) -> None:
+        if self._dirty:
+            state_bits = pickle.dumps(self.state)
+
+            response = self.backend.set_state(state_bits, self._token)
```
its a little bit hard to know how it is going to be used yet, but:

It seems if the node has a dirty state, and the store has also updated state, the node's token will be different from the one on on the store. In this case the response will contain the latest state, and it would override the current node's state. Is it the intended behavior?

cbalioglu replied:
Yes, that is right. Whichever node (that has the same token along with a set of other nodes) acts first, updates the state; other nodes have to retry after fetching the latest state. This ultimately functions as a distributed CAS loop.

tierex commented on torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:
```diff
@@ -243,6 +245,129 @@ def __init__(self) -> None:
+
+            response = self.backend.set_state(state_bits, self._token)
+        else:
+            if self.cache_duration > 0:
```
I don't think we need cache, it would just complicate the rendezvous.

How is it going to be used? - is it going to be periodically invoked in a daemon thread?

cbalioglu replied:
Caching (by default 1 second) actually helps quite a bit if you have separate components in your system that all use the same `_RendezvousStateHolder` instance to retrieve state data (as in `DynamicRendezvousHandler). As they are decoupled from each other, if they frequently query the state, this reduces the amount of traffic.

tierex commented on torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:
```diff
@@ -243,6 +245,129 @@ def __init__(self) -> None:
+        )
+
+        # Filter out the dead nodes.
+        dead_nodes = [
```
Can you explain why do we need to remove dead nodes?

cbalioglu replied:
These are nodes that haven't send a heartbeat for a certain amount of time. We consider such nodes as "dead" and remove them from the rendezvous.

tierex commented on torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:
```diff
@@ -243,6 +245,129 @@ def __init__(self) -> None:
+        dead_nodes = [
+            node
+            for node, last_keep_alive in self.state.last_keep_alives.items()
+            if last_keep_alive < expire_time
```
According to code, we define a node dead if the latest state contains a node that tried to update(or get?) state earlier than `now()-last_keep_alive` . Should we rename "last_keep_alive" to something like "last_heartbeat_time"?

cbalioglu replied:
See the last revision of this PR. I renamed it "keep-alive" to "hearbeat" in most places.

tierex commented on torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:
```diff
@@ -243,6 +245,129 @@ def __init__(self) -> None:
+            if last_keep_alive < expire_time
+        ]
+
+        for dead_node in dead_nodes:
```
If we really need to remove dead_nodes, we probably would like to at least have some logs around what nodes got removed and from what node.

cbalioglu replied:
Makes sense. I will add some logging here.

cbalioglu committed:
Update on "[10/n] [torch/elastic] Introduce _RendezvousStateHolder"


This PR introduces the `_RendezvousStateHolder` type that is responsible for synchronizing the local rendezvous state via the backend with the other nodes in the job.

Differential Revision: [D27892600](https://our.internmc.facebook.com/intern/diff/D27892600/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[10/n] [torch/elastic] Introduce _RendezvousStateHolder"


This PR introduces the `_RendezvousStateHolder` type that is responsible for synchronizing the local rendezvous state via the backend with the other nodes in the job.

Differential Revision: [D27892600](https://our.internmc.facebook.com/intern/diff/D27892600/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[15/n] [torch/elastic] Introduce _RendezvousStateHolder"


This PR introduces the `_RendezvousStateHolder` type that is responsible for synchronizing the local rendezvous state via the backend with the other nodes in the job.

Differential Revision: [D27892600](https://our.internmc.facebook.com/intern/diff/D27892600/)

[ghstack-poisoned]

cbalioglu committed:
Update on "[15/n] [torch/elastic] Introduce _RendezvousStateHolder"


This PR introduces the `_RendezvousStateHolder` type that is responsible for synchronizing the local rendezvous state via the backend with the other nodes in the job.

Differential Revision: [D27892600](https://our.internmc.facebook.com/intern/diff/D27892600/)

[ghstack-poisoned]
